### PR TITLE
feat: CSS ::before/::after content: url() with display:block (fulgur-ai3 Phase 1)

### PR DESCRIPTION
**Pull request by mitsuru**
## Summary

- `::before` / `::after` 疑似要素の `content: url("asset.png")` / `content: image-set(url(...))` を Stylo の computed value から読み出し、AssetBundle 経由で `ImagePageable` として描画する
- `display: block` 疑似要素のみを Phase 1 の対象とする。inline 配置は Phase 2 (`fulgur-n4t`) でフォロー
- 既存 `` 処理を新設の `make_image_pageable` 共通 sizing ヘルパに統合 — 両パスで同じ CSS replaced-element sizing ルール(両方指定/片方+アスペクト補完/intrinsic)を共有

beads issue: [fulgur-ai3](../blob/main/.beads/issues.jsonl) (Phase 1) / `fulgur-n4t` (Phase 2 follow-up)

### 新 API

- `blitz_adapter::extract_pseudo_image_url(&Node) -> Option<String>` — pseudo ノードの computed `content` から single-item `Image` variant の URL を抽出。`image-set(...)` は stylo の `selected_index` をフォローして内部 `Image::Url` に再帰
- `convert::make_image_pageable(data, format, css_w, css_h, opacity, visible) -> ImagePageable` — `` と pseudo-content 両方で共有する sizing / 構築ロジック
- `convert::build_pseudo_image(pseudo_node, parent_width, assets) -> Option<ImagePageable>` — pseudo 単体から ImagePageable を組み立てる。AssetBundle 未登録 / 非 Image content / format 未対応などで silent skip
- `convert::inject_block_pseudo_images(...)` — 親ノードの `PositionedChild` vec に `::before` を先頭挿入 / `::after` を末尾追加する注入ヘルパ
- `convert::has_block_pseudo_image(...)` — inline-root 分岐で BlockPageable ラッピング経路に入るかどうかの cheap pre-check

### 設計上の判断

- **stylo alt text は deferred**: `content: url(...) / "alt"` は stylo の `layout.css.content.alt-text.enabled` pref gate により parser 段階で invalid になり、疑似要素が生成されない。当初 scope に入れていたが Stylo upstream 対応待ちとして Phase 1 から drop。`ImagePageable.alt` フィールドも YAGNI で追加せず
- **通常要素 `content: url()` は out-of-scope**: Blitz 0.2.4 が通常要素の `content: Image` を layout に反映しないため、upstream patch か DOM pre-processing が必要。Phase 2 で対応
- **inline 疑似要素は out-of-scope**: Parley inline layout に非テキスト run を埋め込むには ShapedLine enum 拡張 or U+FFFC sentinel の spike が必要。こちらも Phase 2

### 既知の制約 (Phase 1)

Blitz は text 子を持たない疑似要素に `size=0` の layout を割り当てるため、block pseudo 画像は後続 real child を**押し下げない**。新 example (`examples/pseudo-content-url/`) では `h2` 側に `margin-top: 40px` を置いて手動で space を reserve している。Taffy round-trip 対応 (`fulgur-n4t`) で解消予定。

## Test Plan

- [x] `cargo test --workspace` — 全 pass (244 lib + 11 determinism + 他)
- [x] `cargo fmt --check` — clean
- [x] `cargo clippy --workspace --all-targets -- -D warnings` — clean
- [x] `npx markdownlint-cli2 docs/plans/2026-04-12-pseudo-content-url.md` — 0 error
- [x] manual smoke: `examples/pseudo-content-url/index.pdf` を 2 回生成して byte-identical を確認
- [x] 新規テスト 14 本 (blitz_adapter 3 + convert 10 + examples_determinism 1)
- [ ] CI 上で `pseudo_content_url_example_is_deterministic` が pass することを確認 (フォント差で byte drift しないこと)

## 新規テスト一覧

**`blitz_adapter::tests`** (3):
- `test_extract_pseudo_image_url_simple`
- `test_extract_pseudo_image_url_returns_none_for_string_content`
- `test_extract_pseudo_image_url_image_set`

**`convert::tests`** (10):
- `test_make_image_pageable_both_dimensions`
- `test_make_image_pageable_width_only_uses_intrinsic_aspect`
- `test_make_image_pageable_height_only_uses_intrinsic_aspect`
- `test_make_image_pageable_intrinsic_fallback`
- `test_build_pseudo_image_reads_content_url`
- `test_build_pseudo_image_width_only_uses_intrinsic_aspect`
- `test_build_pseudo_image_missing_asset_returns_none`
- `test_build_pseudo_image_no_assets_returns_none`
- `test_dom_to_pageable_emits_block_pseudo_image`
- `test_dom_to_pageable_inline_pseudo_ignored_phase1`

**`fulgur-cli/tests/examples_determinism.rs`** (1):
- `pseudo_content_url_example_is_deterministic`

---

<a href="https://app.devin.ai/review/mitsuru/fulgur/pull/70" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

## Summary by CodeRabbit

* **新機能**
  * ::before/::after の content: url(...) を持つブロック疑似要素の画像をレンダリングし、疑似要素アイコンが期待通りの位置とサイズで出力されます。
* **テスト**
  * 疑似要素を含む例を2回レンダリングして生成PDFがバイト単位で一致する決定論性テストを追加しました。
* **ドキュメント**
  * 実装計画、例ページ、対応するスタイルシートを追加しました。

### CODE REVIEW

File: crates/fulgur/src/convert.rs [hunk 1]
```diff
@@ -568,6 +572,204 @@ where
...
+    // does not propagate these to `final_layout` for pseudos that lack a text
+    // child, so we must go directly to stylo.
+    let styles = pseudo_node.primary_styles()?;
+    let css_w = resolve_pseudo_size(&styles.clone_width(), parent_width);
+    let css_h = resolve_pseudo_size(&styles.clone_height(), parent_width);
+
+    let (opacity, visible) = extract_opacity_visible(pseudo_node);
+    Some(make_image_pageable(
+        data, format, css_w, css_h, opacity, visible,
+    ))
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert.rs:704]: Fixed in fcca87e. Introduced a ContentBox struct (origin + width + height from compute_content_box()) so: (1) ::before / ::after land at the parent's content-box corner, not the border-box corner; (2) build_pseudo_image takes separate parent_content_width / parent_content_height so height: 50% resolves against height, not width. Added test_build_pseudo_image_height_percent_resolves_against_parent_height to pin the fix.

File: crates/fulgur/src/convert.rs [hunk 2]
```diff
@@ -365,7 +368,8 @@ fn convert_node_inner(
     }
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert.rs:402]: Fixed in fcca87e. The children.is_empty() branch now calls build_block_pseudo_images and enters the BlockPageable path when has_pseudo is true, mirroring the container and inline-root branches. Added regression test test_dom_to_pageable_emits_pseudo_on_childless_element for <div class='icon'></div> with display:block ::before.

File: crates/fulgur/src/convert.rs [hunk 3]
```diff
@@ -270,7 +270,8 @@ fn convert_node_inner(
             }
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert.rs:280]: Fixed in fcca87e. The list-item inline-root body (both styled and previously-unstyled sub-branches) now calls build_block_pseudo_images and wraps the paragraph in a BlockPageable when has_pseudo is true. Added regression test test_dom_to_pageable_emits_pseudo_on_list_item_with_text covering a <li> with inline text and a block ::before. Opacity handling is preserved — ListItemPageable still wraps the whole thing in a single opacity group.

File: crates/fulgur/src/convert.rs [hunk 4]
```diff
@@ -348,9 +376,18 @@ fn convert_node_inner(
...
+        // `<div class="icon"></div>` with `.icon::before { content: url(...) }`
+        // should emit the image. Without this the pseudo is silently dropped.
+        let (before_pseudo, after_pseudo) =
+            build_block_pseudo_images(doc, node, content_box, ctx.assets);
+        let has_pseudo = before_pseudo.is_some() || after_pseudo.is_some();
+        if style.has_visual_style() || style.has_radius() || has_pseudo {
             let (opacity, visible) = extract_opacity_visible(node);
-            let mut block = BlockPageable::with_positioned_children(vec![])
+            let positioned_children =
+                wrap_with_block_pseudo_images(before_pseudo, after_pseudo, content_box, Vec::new());
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert.rs:389]: Fixed in 3d05baf. Added a cheap probe node_has_block_pseudo_image() that walks the pseudo slots looking for a block-display image URL without resolving AssetBundle. collect_positioned_children's zero-size leaf filter now opts out when the probe returns true, so <div class='zero' style='width:0;height:0'></div> with a block ::before reaches convert_node_inner and its empty-children branch emits the image. Regression test: test_dom_to_pageable_emits_pseudo_on_zero_size_block_leaf.

Scope note: inline <span class='icon'></span> inside a paragraph is a different edge case — the span participates in Parley inline layout rather than reaching collect_positioned_children as a standalone child. That path requires Phase 2 inline-flow handling and is tracked in fulgur-n4t.

**Comment by mitsuru** [comment]
@coderabbitai resolve
